src/script/interpreter.cpp

```diff
@@ -1511,17 +1512,54 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
     }
 
     // Run the script interpreter.
-    if (!EvalScript(stack, scriptPubKey, flags, checker, sigversion, serror)) return false;
+    if (!EvalScript(stack, script, flags, checker, sigversion, serror)) return false;
 
     // Scripts inside witness implicitly require cleanstack behaviour
     if (stack.size() != 1) return set_error(serror, SCRIPT_ERR_CLEANSTACK);
     if (!CastToBool(stack.back())) return set_error(serror, SCRIPT_ERR_EVAL_FALSE);
     return true;
 }
 
-static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion, const std::vector<unsigned char>& program, unsigned int flags, const BaseSignatureChecker& checker, ScriptError* serror)
+/** Consensus-critical analogue of CScript::IsPayToScriptHash. */
+static bool IsPayToScriptHash(const Span<const unsigned char>& script)
+{
+    return script.size() == 23 &&
+           script[0] == OP_HASH160 &&
+           script[1] == 0x14 &&
+           script[22] == OP_EQUAL;
+}
+
+/** Consensus-critical analogue of CScript::IsWitnessProgram. */
+static bool IsWitnessProgram(const Span<const unsigned char>& script, int& version, Span<const unsigned char>& program)
+{
+    if (script.size() < 4 || script.size() > 42) return false;
+    if (script[0] != OP_0 && (script[0] < OP_1 || script[0] > OP_16)) return false;
+    if (std::size_t(script[1] + 2) == script.size()) {
+        version = CScript::DecodeOP_N((opcodetype)script[0]);
+        program = script.subspan(2);
+        return true;
+    }
+    return false;
+}
+
+/** Consensus-critical analogue of CScript::IsPushOnly. */
+static bool IsPushOnly(const Span<const unsigned char>& script)
```

Comment:
Any reason to make this private? `static` makes it harder to test and also the "other" version of IsPushOnly can't call this to avoid code duplication.